### ISSUE TITLE
Compilation error error: call of overloaded ‘Package(const nlohmann::basic_json<>&)’ is ambiguous

### ISSUE THREAD

**Issue by inglor** [falconindy/auracle#53]
Trying to compile latest master (5e7bcad)

```
inglor@tiamat /tmp$ git clone https://github.com/falconindy/auracle
Cloning into 'auracle'...
remote: Enumerating objects: 308, done.
remote: Counting objects: 100% (308/308), done.
remote: Compressing objects: 100% (152/152), done.
remote: Total 1713 (delta 161), reused 250 (delta 148), pack-reused 1405
Receiving objects: 100% (1713/1713), 682.90 KiB | 1.78 MiB/s, done.
Resolving deltas: 100% (1072/1072), done.
inglor@tiamat /tmp$ cd auracle
inglor@tiamat /tmp/auracle$ meson build --prefix=/usr --buildtype=plain --default-library=static
The Meson build system
Version: 0.50.1
Source dir: /tmp/auracle
Build dir: /tmp/auracle/build
Build type: native build
Project name: auracle
Project version: 0
Native C++ compiler: ccache c++ (gcc 8.3.0 "c++ (GCC) 8.3.0")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Configuring config.h using configuration
Compiler for C++ supports arguments -ffunction-sections: YES
Compiler for C++ supports arguments -fdata-sections: YES
Compiler for C++ supports link arguments -Wl,--gc-sections: YES
Found pkg-config: /bin/pkg-config (1.6.1)
Dependency libalpm found: YES 11.0.3
Dependency libarchive found: YES 3.3.3
Dependency libcurl found: YES 7.64.1
Dependency libsystemd found: YES 242
Dependency GTest found: YES 1.8.1
Dependency GMock found: YES 1.8.1
Library stdc++fs found: YES
Dependency fmt found: YES 5.3.0
Program pod2man found: YES (/usr/bin/core_perl/pod2man)
Build targets in project: 12
Found ninja-1.9.0 at /bin/ninja
inglor@tiamat /tmp/auracle$ ninja -C build
ninja: Entering directory `build'
[12/27] Compiling C++ object 'aur@sta/src_aur_response.cc.o'.
FAILED: aur@sta/src_aur_response.cc.o
ccache c++ -Iaur@sta -I. -I.. -I../third_party/nlohmann_json -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++17 -include config.h -ffunction-sections -fdata-sections -fPIC -MD -MQ 'aur@sta/src_aur_response.cc.o' -MF 'aur@sta/src_aur_response.cc.o.d' -o 'aur@sta/src_aur_response.cc.o' -c ../src/aur/response.cc
In file included from /usr/include/c++/8.3.0/vector:62,
                 from ../src/aur/package.hh:6,
                 from ../src/aur/response.hh:4,
                 from ../src/aur/response.cc:1:
/usr/include/c++/8.3.0/bits/stl_construct.h: In instantiation of ‘void std::_Construct(_T1*, _Args&& ...) [with _T1 = aur::Package; _Args = {const nlohmann::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long int, long unsigned int, double, std::allocator, nlohmann::adl_serializer>&}]’:
/usr/include/c++/8.3.0/bits/stl_uninitialized.h:83:18:   required from ‘static _ForwardIterator std::__uninitialized_copy<_TrivialValueTypes>::__uninit_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _ForwardIterator = aur::Package*; bool _TrivialValueTypes = false]’
/usr/include/c++/8.3.0/bits/stl_uninitialized.h:134:15:   required from ‘_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _ForwardIterator = aur::Package*]’
/usr/include/c++/8.3.0/bits/stl_uninitialized.h:289:37:   required from ‘_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _ForwardIterator = aur::Package*; _Tp = aur::Package]’
/usr/include/c++/8.3.0/bits/stl_vector.h:1469:33:   required from ‘void std::vector<_Tp, _Alloc>::_M_range_initialize(_ForwardIterator, _ForwardIterator, std::forward_iterator_tag) [with _ForwardIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _Tp = aur::Package; _Alloc = std::allocator<aur::Package>]’
/usr/include/c++/8.3.0/bits/stl_vector.h:1437:4:   required from ‘void std::vector<_Tp, _Alloc>::_M_initialize_dispatch(_InputIterator, _InputIterator, std::__false_type) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _Tp = aur::Package; _Alloc = std::allocator<aur::Package>]’
/usr/include/c++/8.3.0/bits/stl_vector.h:546:4:   required from ‘std::vector<_Tp, _Alloc>::vector(_InputIterator, _InputIterator, const allocator_type&) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; <template-parameter-2-2> = void; _Tp = aur::Package; _Alloc = std::allocator<aur::Package>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<aur::Package>]’
../src/aur/json_internal.hh:33:12:   required from ‘void aur::from_json(const json&, std::vector<CharType>&) [with T = aur::Package; nlohmann::json = nlohmann::basic_json<>]’
../src/aur/json_internal.hh:42:62:   required from ‘aur::ValueCallback<OutputType> aur::MakeValueCallback(FieldType OutputType::*) [with OutputType = aur::RpcResponse; FieldType = std::vector<aur::Package>; aur::ValueCallback<OutputType> = std::function<void(const nlohmann::basic_json<>&, aur::RpcResponse&)>]’
../src/aur/response.cc:14:62:   required from here
/usr/include/c++/8.3.0/bits/stl_construct.h:75:7: error: call of overloaded ‘Package(const nlohmann::basic_json<>&)’ is ambiguous
     { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/aur/response.hh:4,
                 from ../src/aur/response.cc:1:
../src/aur/package.hh:28:12: note: candidate: ‘aur::Package::Package(const string&)’
   explicit Package(const std::string& json_bytes);
            ^~~~~~~
../src/aur/package.hh:26:8: note: candidate: ‘aur::Package::Package(const aur::Package&)’
 struct Package {
        ^~~~~~~
../src/aur/package.hh:26:8: note: candidate: ‘aur::Package::Package(aur::Package&&)’
[19/27] Compiling C++ object 'auracle@sta/src_auracle_auracle.cc.o'.
ninja: build stopped: subcommand failed.
```

**Comment by falconindy** [comment]
gcc 9.1.0 = works
clang++ 8.0.0 = works

it's a gcc 8 bug.

**Comment by Louson** [comment]
How do you build with clang ? Is there an argument to give to ninja ?
Is there a way to give it straight through makepkg ?

**Comment by falconindy** [comment]
You can export CXX=clang++ as an environment var.

**Comment by Louson** [comment]
Thanks, I missed that I had to do it before the command `meson build`

**Comment by falconindy** [comment]
You can do it with makepkg as well,

`makepkg -- CXX=clang++`

Should work.

**Comment by Emi1305** [comment]
> You can do it with makepkg as well,
> 
> `makepkg -- CXX=clang++`
> 
> Should work.

Can't make this work, for some reason it sticks to gcc

```
makepkg -- CXX=clang++
==> Making package: auracle-git r254.5e7bcad-1 (Mon 27 May 2019 05:29:33 PM -03)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Updating auracle git repo...
Fetching origin
==> Validating source files with sha256sums...
    auracle ... Skipped
==> Extracting sources...
  -> Creating working copy of auracle git repo...
Reset branch 'makepkg'
==> Starting pkgver()...
==> Removing existing $pkgdir/ directory...
==> Starting build()...
ninja: Entering directory `build'
[0/1] Regenerating build files.
The Meson build system
Version: 0.50.1
Source dir: /opt/auracle-git/src/auracle
Build dir: /opt/auracle-git/src/auracle/build
Build type: native build
Project name: auracle
Project version: 0
Native C++ compiler: c++ (gcc 8.3.0 "c++ (GCC) 8.3.0")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Configuring config.h using configuration
Compiler for C++ supports arguments -ffunction-sections: YES
Compiler for C++ supports arguments -fdata-sections: YES
Compiler for C++ supports link arguments -Wl,--gc-sections: YES
Dependency libalpm found: YES (cached)
Dependency libarchive found: YES (cached)
Dependency libcurl found: YES (cached)
Dependency libsystemd found: YES (cached)
Dependency gtest found: YES (cached)
Dependency gmock found: YES (cached)
Library stdc++fs found: YES
Found pkg-config: /usr/bin/pkg-config (1.6.1)
Found CMake: /usr/bin/cmake (3.14.4)
Dependency fmt found: NO (tried pkgconfig and cmake)

|Executing subproject fmt 
|
|Project name: fmt
|Project version: 5.3.0
|Native C++ compiler: c++ (gcc 8.3.0 "c++ (GCC) 8.3.0")
|Build targets in project: 1
|Subproject fmt finished.

Program pod2man found: YES (/usr/bin/core_perl/pod2man)
Build targets in project: 13
Found ninja-1.9.0 at /usr/bin/ninja
[1/8] Compiling C++ object 'aur@sta/src_aur_response.cc.o'.
FAILED: aur@sta/src_aur_response.cc.o 
c++ -Iaur@sta -I. -I.. -I../third_party/nlohmann_json -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++17 -include config.h -ffunction-sections -fdata-sections -march=x86-64 -mtune=generic -O2 -fno-plt -D_FORTIFY_SOURCE=2 -fPIC -MD -MQ 'aur@sta/src_aur_response.cc.o' -MF 'aur@sta/src_aur_response.cc.o.d' -o 'aur@sta/src_aur_response.cc.o' -c ../src/aur/response.cc
In file included from /usr/include/c++/8.3.0/vector:62,
                 from ../src/aur/package.hh:6,
                 from ../src/aur/response.hh:4,
                 from ../src/aur/response.cc:1:
/usr/include/c++/8.3.0/bits/stl_construct.h: In instantiation of ‘void std::_Construct(_T1*, _Args&& ...) [with _T1 = aur::Package; _Args = {const nlohmann::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long int, long unsigned int, double, std::allocator, nlohmann::adl_serializer>&}]’:
/usr/include/c++/8.3.0/bits/stl_uninitialized.h:83:18:   required from ‘static _ForwardIterator std::__uninitialized_copy<_TrivialValueTypes>::__uninit_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _ForwardIterator = aur::Package*; bool _TrivialValueTypes = false]’
/usr/include/c++/8.3.0/bits/stl_uninitialized.h:134:15:   required from ‘_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _ForwardIterator = aur::Package*]’
/usr/include/c++/8.3.0/bits/stl_uninitialized.h:289:37:   required from ‘_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _ForwardIterator = aur::Package*; _Tp = aur::Package]’
/usr/include/c++/8.3.0/bits/stl_vector.h:1469:33:   required from ‘void std::vector<_Tp, _Alloc>::_M_range_initialize(_ForwardIterator, _ForwardIterator, std::forward_iterator_tag) [with _ForwardIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _Tp = aur::Package; _Alloc = std::allocator<aur::Package>]’
/usr/include/c++/8.3.0/bits/stl_vector.h:1437:4:   required from ‘void std::vector<_Tp, _Alloc>::_M_initialize_dispatch(_InputIterator, _InputIterator, std::__false_type) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; _Tp = aur::Package; _Alloc = std::allocator<aur::Package>]’
/usr/include/c++/8.3.0/bits/stl_vector.h:546:4:   required from ‘std::vector<_Tp, _Alloc>::vector(_InputIterator, _InputIterator, const allocator_type&) [with _InputIterator = nlohmann::detail::iter_impl<const nlohmann::basic_json<> >; <template-parameter-2-2> = void; _Tp = aur::Package; _Alloc = std::allocator<aur::Package>; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<aur::Package>]’
../src/aur/json_internal.hh:33:12:   required from ‘void aur::from_json(const json&, std::vector<CharType>&) [with T = aur::Package; nlohmann::json = nlohmann::basic_json<>]’
../src/aur/json_internal.hh:42:62:   required from ‘aur::ValueCallback<OutputType> aur::MakeValueCallback(FieldType OutputType::*) [with OutputType = aur::RpcResponse; FieldType = std::vector<aur::Package>; aur::ValueCallback<OutputType> = std::function<void(const nlohmann::basic_json<>&, aur::RpcResponse&)>]’
../src/aur/response.cc:14:62:   required from here
/usr/include/c++/8.3.0/bits/stl_construct.h:75:7: error: call of overloaded ‘Package(const nlohmann::basic_json<>&)’ is ambiguous
     { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/aur/response.hh:4,
                 from ../src/aur/response.cc:1:
../src/aur/package.hh:28:12: note: candidate: ‘aur::Package::Package(const string&)’
   explicit Package(const std::string& json_bytes);
            ^~~~~~~
../src/aur/package.hh:26:8: note: candidate: ‘aur::Package::Package(const aur::Package&)’
 struct Package {
        ^~~~~~~
../src/aur/package.hh:26:8: note: candidate: ‘aur::Package::Package(aur::Package&&)’
ninja: build stopped: subcommand failed.
==> ERROR: A failure occurred in build().
    Aborting...

```

Exporting `CC` and `CXX` before calling makepkg and editing `/etc/makepkg.conf` didn't work either. Any ideas?

**Comment by falconindy** [comment]
Works for me... you need to start with a clean build dir.

**Comment by Emi1305** [comment]
Oh, `makepkg -C CXX=clang++` did the trick, thank you so much

**Comment by inglor** [comment]
`gcc 9` was [removed from testing] but `clang 8` as suggested works fine. I'll keep this open until `gcc` comes on `core`.

[removed from testing]: https://lists.archlinux.org/pipermail/arch-dev-public/2019-May/029589.html

**Comment by falconindy** [comment]
Should be fixed with 543585516b382e7db8a86433a2251e548c569547

**Comment by inglor** [comment]
> Should be fixed with [5435855](https://github.com/falconindy/auracle/commit/543585516b382e7db8a86433a2251e548c569547)

It is - thanks.
